flipt-client-go/evaluation.go

```diff
@@ -128,151 +124,663 @@ func WithJWTAuthentication(token string) clientOption {
+				return fmt.Errorf("failed to update engine: %w", err)
+			}
+
+			// dont defer in loop to avoid stack overflow
+			e.deallocFunc.Call(ctx, uint64(pmPtr[0]), uint64(len(s.payload)))
+		}
 	}
+}
 
-	flags := C.list_flags(e.engine)
-	defer C.destroy_string(flags)
+// startPolling starts the background polling for the given update interval.
+func (e *EvaluationClient) startPolling(ctx context.Context) {
+	ticker := time.NewTicker(e.updateInterval)
+	defer ticker.Stop()
+
+	// create a new context for the polling so any timeouts
+	// are independent of the client context
+	fctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			e.mu.RLock()
+			etag := e.etag
+			e.mu.RUnlock()
+
+			snapshot, err := e.fetch(fctx, etag)
+			if err != nil {
+				e.errChan <- err
+				continue
+			}
+
+			e.snapshotChan <- snapshot
+		}
+	}
+}
 
-	b := C.GoBytes(unsafe.Pointer(flags), (C.int)(C.strlen(flags)))
+// fetch fetches the snapshot for the given namespace.
+func (e *EvaluationClient) fetch(ctx context.Context, etag string) (snapshot, error) {
+	url := fmt.Sprintf("%s/internal/v1/evaluation/snapshot/namespace/%s", e.url, e.namespace)
+	if e.ref != "" {
+		url = fmt.Sprintf("%s?reference=%s", url, e.ref)
+	}
 
-	var fl *ListFlagsResult
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return snapshot{}, fmt.Errorf("failed to create request: %w", err)
+	}
 
-	if err := json.Unmarshal(b, &fl); err != nil {
-		return nil, err
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("x-flipt-accept-server-version", "1.47.0")
+
+	if e.authentication != nil {
+		switch auth := e.authentication.(type) {
+		case clientTokenAuthentication:
+			req.Header.Set("Authorization", "Bearer "+auth.Token)
+		case jwtAuthentication:
+			req.Header.Set("Authorization", "JWT "+auth.Token)
+		}
+	}
+
+	if etag != "" {
+		req.Header.Set("If-None-Match", etag)
+	}
+
+	resp, err := e.httpClient.Do(req)
+	if err != nil {
+		return snapshot{}, fetchError(fmt.Errorf("failed to fetch snapshot: %w", err))
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNotModified {
+		return snapshot{}, nil
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return snapshot{}, fetchError(fmt.Errorf("unexpected status code: %d", resp.StatusCode))
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return snapshot{}, fetchError(fmt.Errorf("failed to read response body: %w", err))
+	}
+
+	etag = resp.Header.Get("ETag")
+	return snapshot{payload: body, etag: etag}, nil
+}
+
+type streamChunk struct {
+	Result streamResult
+}
+
+type streamResult struct {
+	Namespaces map[string]json.RawMessage
+}
+
+// startStreaming starts the background streaming.
+// Note: currently any errors cause this method to exit. We still need to implement retries for trying to reconnect.
+func (e *EvaluationClient) startStreaming(ctx context.Context) {
```

Comment:
Probably there should be some reconnect logic. It will work fine probably to the first connection/timeout error and then client will not get any updates from Flipt (for example, the restart will kill all streaming for the clients).  I don't know if I can emulate this with current open-source version. I just make a wild guess here.

Reply:
yeah you are right again! i was thinking of doing the retry/reconnect logic in another PR before release as this one is already quite large. I have it as a todo on the PR  description though.